crates/ic-ingress-validator-util/src/delegation_validator.rs

```diff
@@ -0,0 +1,382 @@
+                f,
+                "The user id {} does not match the public key {}",
+                user_id,
+                hex::encode(pubkey)
+            ),
+            InvalidSignature(err) => write!(f, "Invalid signature: {}", err),
+            InvalidDelegation(err) => write!(f, "Invalid delegation: {}", err),
+            MissingSignature(user_id) => write!(f, "Missing signature from user: {}", user_id),
+            AnonymousSignatureNotAllowed => {
+                write!(f, "Signature is not allowed for the anonymous user")
+            }
+            CanisterNotInDelegationTargets(canister_id) => write!(
+                f,
+                "Canister {} is not one of the delegation targets",
+                canister_id
+            ),
+            TooManyPathsError { length, maximum } => write!(
+                f,
+                "Too many paths in read state request: got {} paths, but at most {} are allowed",
+                length, maximum
+            ),
+            PathTooLongError { length, maximum } => write!(
+                f,
+                "At least one path in read state request is too deep: got {} labels, but at most {} are allowed",
+                length, maximum
+            ),
+            NonceTooBigError { num_bytes: length, maximum } => write!(
+                f,
+                "Nonce in request is too big: got {} bytes, but at most {} are allowed",
+                length, maximum
+            ),
+        }
+    }
+}
+
+/// Error in verifying the signature or authentication part of a request.
+#[derive(Debug, PartialEq, thiserror::Error)]
+pub enum AuthenticationError {
+    InvalidBasicSignature(CryptoError),
+    InvalidCanisterSignature(String),
+    InvalidPublicKey(CryptoError),
+    WebAuthnError(String),
+    DelegationTargetError(String),
+    DelegationTooLongError { length: usize, maximum: usize },
+    DelegationContainsCyclesError { public_key: Vec<u8> },
+}
+
+impl fmt::Display for AuthenticationError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            InvalidBasicSignature(err) => write!(f, "Invalid basic signature: {}", err),
+            InvalidCanisterSignature(err) => write!(f, "Invalid canister signature: {}", err),
+            InvalidPublicKey(err) => write!(f, "Invalid public key: {}", err),
+            WebAuthnError(msg) => write!(f, "{}", msg),
+            DelegationTargetError(msg) => write!(f, "{}", msg),
+            DelegationTooLongError { length, maximum } => write!(
+                f,
+                "Chain of delegations is too long: got {} delegations, but at most {} are allowed",
+                length, maximum
+            ),
+            DelegationContainsCyclesError { public_key } => write!(
+                f,
+                "Chain of delegations contains at least one cycle: first repeating public key encountered {}",
+                hex::encode(public_key)
```

Comment:
especially if these are ultimately ed25519 public keys it might be more conventional to base58 code this when displaying